File: jvm-experiments-commons/src/main/java/kr/kth/commons/timeperiod/YearAndHalfYear.java
package kr.kth.commons.timeperiod;

import com.google.common.base.Objects;
import kr.kth.commons.base.ValueObjectBase;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

/**
 * kr.kth.commons.timeperiod.YearAndHalfYear
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 23.
 */
@Slf4j
public class YearAndHalfYear extends ValueObjectBase implements Comparable<YearAndHalfYear> {

	private static final long serialVersionUID = 4272394545140829436L;

	@Getter @Setter
	private Integer year;
	@Getter @Setter
	HalfYearKind halfyear;

	public YearAndHalfYear() {}

	public YearAndHalfYear(Integer year, HalfYearKind halfyear) {
		this.year = year;
		this.halfyear = halfyear;
	}

	public YearAndHalfYear(Integer year, Integer halfyear) {
		this.year = year;
		if (halfyear != null) {
			switch (halfyear) {
				case 1:
					this.halfyear = HalfYearKind.First;
					break;
				case 2:
					this.halfyear = HalfYearKind.Second;
					break;
			}
		}
	}

	@Override
	public int compareTo(YearAndHalfYear o) {
		return hashCode() - o.hashCode();
	}

	@Override
	public int hashCode() {
		return ((year != null) ? year : 0) * 100 + ((halfyear != null) ? halfyear.getValue() : 0);
	}

	@Override
	protected Objects.ToStringHelper buildStringHelper() {
		return super.buildStringHelper()
		            .add("year", year)
		            .add("halfyear", halfyear);
	}
}

File: jvm-experiments-commons/src/main/scala/kr/kth/commons/timeperiod/YearAndHalfyear.scala
package kr.kth.commons.timeperiod

import beans.BeanProperty
import clock.ClockProxy
import kr.kth.commons.base.ValueObjectBase
import org.joda.time.DateTime

/**
 * 년(Year), 반기(HalfYear) 정보를 가지는 Value Object 입니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 26
 */
class YearAndHalfyear(@BeanProperty val year: Int,
                      @BeanProperty val halfyear: HalfYearKind = HalfYearKind.First)
    extends ValueObjectBase with Comparable[YearAndHalfyear] {

    def this(year: Int, halfyear: Int = 1) = this(year, HalfYearKind.valueOf(halfyear))

    def compareTo(other: YearAndHalfyear) = hashCode compareTo other.hashCode

    override def hashCode: Int = year * 100 + halfyear.toInt

    protected override def buildStringHelper =
        super.buildStringHelper
            .add("year", year)
            .add("halfyear", halfyear)
}

object YearAndHalfyear {

    def apply(): YearAndHalfyear = apply(ClockProxy.getClock.today())

    def apply(today: DateTime): YearAndHalfyear = {
        val today = ClockProxy.getClock.today()
        new YearAndHalfyear(today.getYear, Times.halfyearOf(today))
    }
}
